### PR TITLE
add documentation to coordination and clean up unused functions

### PR DESCRIPTION
**Pull request by qema**
- added configuration options to Coordination constructor, with documentation
- add documentation for class
- add coordination_attribute_name parameter to constructor for custom metadata keys
- remove/deprecate precompute() workflow. directly use fit() instead
- deprecate utterance_thresh_func(). use the various speaker, target and utterance count-based thresholds instead

### CODE REVIEW

**Comment by calebchiam** [comment]
`scores_over_utterances()` still takes in the `split_by_attribs`, `speaker_attribs`, `target_attribs` arguments -- just wanted to confirm that these are still relevant and check whether they need to be included in the constructor. Even if they should be kept, they should be refactored to lambda selectors (see the `selector` argument in other Transformers).

How do the new changes work with the Coordination example notebook?

**Comment by qema** [comment]
- These arguments to `scores_over_utterances()` shouldn't be needed in the constructor since they don't apply to the raw pairwise coordination scores that `transform()` gets. After looking into it, I believe they function somewhat differently from selectors -- they are each a list of attributes which can be used to split the set of speakers/targets into subgroups for use in aggregating coordination scores. For example, the set of Supreme Court justices can be split into the set of "favorable" vs "unfavorable" justices by splitting on the "is-favorable" attribute. Whereas I believe selectors are used to take a subset of items (as opposed to partitioning the set?)
- Currently, `score()` (and associated `score_report()`) are also front-facing convenience methods that aggregate the coordination scores (so instead of getting pairwise scores between all speaking pairs, we might have a score from all lawyers to all justices). I could rename `score()` to `summarize()` (and fold it together with `score_report()`) to be more in line with the other Transformers if that makes sense to you (not sure if the output has to be a Pandas DataFrame?)

Confirmed that the example notebook still functions as before with the changes.

**Comment by calebchiam** [comment]
Hey @qema, thanks for the response. Some further questions and answers:

- So it sounds like `transform()` only supports raw pairwise coordination scores. I'm not sure if there are obstacles to this, but would it perhaps make sense for `transform()` to be more 'general' in that it should be able to support splitting speakers/targets into subgroups as well? This is to keep in line with the Transformer interface, where the main methods to make use of the Transformer should be `fit`, `transform`, and `summarize`, with other methods being of secondary importance. In that case, it would perhaps make sense for the arguments to be available in `transform()`.
- For selectors, what I had in mind was that you could specify the two groups using two selectors, e.g. a lambda for speaker.meta['is-favorable'] = 'favourable' and another for speaker.meta['is-favorable'] = 'unfavourable'. In this case, I agree the attribute approach might appear simpler. But for experimentation and when you might want to split into different groups on the fly, the attribute approach might be more tedious to work with than the more versatile lambda (which would let you define the two groups in very flexible and dynamic ways). This is the approach we've gone with in other similar transformers, e.g. FightingWords, where we had the same design choice to make.
- I think the `summarize()` implementation you suggested makes perfect sense. (Output can indeed be anything we'd like.)

**Review by calebchiam — CHANGES_REQUESTED** [review]

**Comment by qema** [comment]
- My feeling is that the extra functionality involving subgroups falls under "ways to aggregate the raw scores in a more convenient way", in which case I can make all of this functionality available under `summarize()`. Yes, ideally we should have all functionality be accessible via fit, transform and summarize.
- I had the impression of arbitrary lambda selectors being a bit tricky in the coordination case but I'll take another look (using FightingWords for reference).

Will work on these changes tomorrow but may need to finish later in the week if that works.

File: convokit/coordination/coordination.py [hunk 1]
```diff
@@ -5,26 +5,72 @@
...
-    def __init__(self, **thresh):
-        self.thresh = thresh
+    def __init__(self, coordination_attribute_name: str = "coord", 
+        speaker_thresh: int = 0, target_thresh: int = 3,
+        utterances_thresh: int = 0, speaker_thresh_indiv: int = 0,
+        target_thresh_indiv: int = 0, utterances_thresh_indiv: int = 0,
+        utterance_thresh_func: Optional[Callable[[Tuple[Utterance, Utterance]], bool]] = None):
+        if utterance_thresh_func is not None:
+            deprecation("Coordination's utterance_thresh_func parameter",
+                "speaker, target and utterance threshold parameters")
```
[the remark this answers is not in the document]
Comment (qema) [reply convokit/coordination/coordination.py:49]: scores_over_utterances() does, but I think this argument was more for when I was prototyping and has been superseded by the simpler/numerous numerical thresholds

File: convokit/coordination/coordination.py [hunk 2]
```diff
@@ -33,26 +79,27 @@ def transform(self, corpus: Corpus) -> Corpus:
...
 
-        pair_scores = self.pairwise_scores(corpus, corpus.speaking_pairs(), **self.thresh)
+        pair_scores = self.pairwise_scores(corpus, corpus.speaking_pairs(),
+            speaker_thresh=self.speaker_thresh,
+            target_thresh=self.target_thresh,
+            utterances_thresh=self.utterances_thresh,
+            speaker_thresh_indiv=self.speaker_thresh_indiv,
+            target_thresh_indiv=self.target_thresh_indiv,
+            utterances_thresh_indiv=self.utterances_thresh_indiv,
+            utterance_thresh_func=self.utterance_thresh_func)
```
[the remark this answers is not in the document]
Comment (qema) [reply convokit/coordination/coordination.py:89]: yes sounds great to move these to `summarize()`

**Comment by calebchiam** [comment]
- Ah, I see. Yes, that makes sense to me. It would be good to maintain this distinction between the 'different ways of aggregating the raw scores' in the code by keeping them as separate _private methods, and then have summarize optionally call and return each of them.

Sounds good -- thanks again @qema!

**Comment by qema** [comment]
Makes sense!

**Comment by qema** [comment]
- folded score() and score_report() into new summarize() method. summarize() has a flag summary_report that, when true, generates a dictionary of global coordination statistics (presenting the info previously obtained through score_report() in a slightly more accessible format). When false, summarize() generates a dictionary of aggregated scores from each speaker to the entire set of speakers in the opposite group.
- updated the documentation and example notebook to reflect the new usage

Re changing the split_by_attrib arguments to selectors, my current impression is that the coordination case differs from FightingWords in that for coordination, there may be arbitrarily many "classes" (groups of speakers) induced by a given attribute (whereas there are always two classes for FightingWords). Furthermore, the current formulation assumes that each speaker belongs to exactly one class (whereas defining the classes via selectors may result in overlapping classes). I think these difficulties could be overcome, but in general it may be more natural in the coordination case for the splits to define a "partition" of the set of speakers, so that the summarize method is seen as giving a summary at different levels of fine-grainedness? On the fence about this, would be interested in your thoughts. One way to add additional flexibility could be to add an arbitrary "partitioning" function that takes in a speaker and outputs a partition class, so that the subdivisions do not have to follow the metadata attributes.

(Will make a separate PR for the supreme court v2 related changes.)

File: convokit/coordination/coordination.py [hunk 3]
```diff
@@ -101,16 +101,29 @@ def fit_transform(self, corpus: Corpus, y=None) -> Corpus:
...
               group: Collection[Union[Speaker, str]], focus: str = "speakers",
-              speaker_thresh: int = 0, target_thresh: int = 3,
-              utterances_thresh: int = 0, speaker_thresh_indiv: int = 0,
-              target_thresh_indiv: int = 0,
-              utterances_thresh_indiv: int = 0,
+              summary_report: bool = False,
+              speaker_thresh: int = -1, target_thresh: int = -1,
+              utterances_thresh: int = -1, speaker_thresh_indiv: int = -1,
+              target_thresh_indiv: int = -1,
+              utterances_thresh_indiv: int = -1,
```
Comment (calebchiam) [inline convokit/coordination/coordination.py:110]: Could we use the Optional[int] style here?

**Review by calebchiam — COMMENTED** [review]
Just one code suggestion, otherwise it looks good! 👍 

Regarding your comments, I'm not sure whether that's an actual distinction between FW and Coord. One point of clarification is that in the FW case, it's completely possible to define overlapping classes, but this is just treated as failure on the part of the user (and easy enough to catch in the code if we wanted to -- though we currently do not). 

I was also under the impression that there are only ever two classes of speakers in Coordination, so if that's not the case, then this might be more trouble than it's worth. I'm not familiar enough with Coordination to discuss the merits of your proposal at that level though... @cristiandnm, could you weigh in?

File: convokit/coordination/coordination.py (hunk 3 above)
Comment (qema) [reply convokit/coordination/coordination.py]: updated!

**Comment by qema** [comment]
Ah interesting, maybe the fact that the partitions cannot overlap is not too important then.

To clarify the setup with splitting, summarize() always computes coordination between two groups of speakers. The splitting determines what level of resolution the summary statistics should be within these groups. So for example, if we compute coordination between justices and lawyers, and split by whether a justice is favorable, we will get coordination between favorable justices and layers, and unfavorable justices and lawyers. There are also applications with more than two splits, such as splitting lawyers by case as in the paper (so we treat the coordination from justice A on case 1 against all lawyers, the coordination from justice A on case 2 against all lawyers, ... as separate scores before aggregating).

File: convokit/coordination/coordination.py (hunk 3 above)
Comment (calebchiam) [reply convokit/coordination/coordination.py]: Ah, but this also means the default should then be None instead of -1.

**Comment by calebchiam** [comment]
Ah, thanks for the concise explanation. I don't see a neat way for replacing split_by_attribs with selectors, but the other parts seem feasible. Perhaps this approach would make sense?

- Two selectors for defining the two groups of speakers
- Two more selectors for filtering utterances: speaker_utterance_selector and target_utterance_selector; this will replace _utterance_has_attribs, speaker_attribs, target_attribs
- Keep split_by_attribs as it currently is

File: convokit/coordination/coordination.py (hunk 3 above)
Comment (qema) [reply convokit/coordination/coordination.py]: fixed

**Comment by qema** [comment]
Sounds great to me. Will make those changes tomorrow.

File: convokit/coordination/coordination.py [hunk 4]
```diff
@@ -169,8 +179,18 @@ def summarize(self, corpus: Corpus, speakers: Collection[Union[Speaker, str]],
...
+        else:
+            deprecation("Coordination's speaker_attribs parameter",
+                'speaker_utterance_selector')
+            raise ValueError
+        if target_attribs is None:
+            target_attribs = dict()
+        else:
+            deprecation("Coordination's target_attribs parameter",
+                'target_utterance_selector')
+            raise ValueError
```
Comment (calebchiam) [inline convokit/coordination/coordination.py:193]: Will this break backwards compatibility?

Comment (qema) [reply convokit/coordination/coordination.py]: just added a few lines to reproduce the old behavior when these arguments are supplied

**Comment by calebchiam** [comment]
Looks great! Thank you for the hard work, merging now.

**Pull request merged**